### PR TITLE
Universal listings - add filters for page type and date updated

### PR DESCRIPTION
**Pull request by gasman**

### CODE REVIEW

File: wagtail/admin/views/pages/listing.py
Comment (laymonage) [inline wagtail/admin/views/pages/listing.py:365]: Good catch!

File: wagtail/admin/views/pages/listing.py
Comment (laymonage) [inline wagtail/admin/views/pages/listing.py:54]: I was curious why the `latest_revision_created_at` works without specifying it here, but it looks like you only need to specify the fields that you want to be automatically generated (like `ModelForm`).

However, we can't leave out the `fields` definition, so this should probably be

```suggestion
        fields = []
```

or include both fields if we want to be explicit. Maybe add a comment to explain?

File: wagtail/admin/views/pages/listing.py
Comment (laymonage) [inline wagtail/admin/views/pages/listing.py:41]: Unfortunately the `search_all` parameter isn't retained upon applying the filters.

I managed to hack it to work by adding the following:

```suggestion
from django.forms import HiddenInput

class PageFilterSet(WagtailFilterSet):
    def noop_filter(self, qs, name, value):
        """
        A filter method that does not affect the queryset, useful for retaining
        query parameters that are handled outside of the filterset.
        """
        return qs

    search_all = django_filters.BooleanFilter(widget=HiddenInput, method="noop_filter")
```

If we want this, I think the `noop_filter` method would be a useful addition to `WagtailFilterSet`.

**Review by laymonage — COMMENTED** [review]
The filters work very well, thank you! Just a few suggestions for the UI and retaining the "Search all" parameter.

File: wagtail/admin/templates/wagtailadmin/pages/page_listing_header.html
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/pages/page_listing_header.html:47]: Ah, just noticed something: make sure the include is outside the `<form>` element for the search, because nested `<form>` isn't valid HTML5.

Comment (gasman) [reply wagtail/admin/templates/wagtailadmin/pages/page_listing_header.html:47]: Nice one! I like the `{% dialog %}` approach, especially as it means we can drop the temporary CSS hacks I added...

I've decided to ditch the shared/filters.html include in favour of defining them directly in page_listing_header.html - it seems there's more stuff in there that we don't want (such as the duplicate `<form>` element) than we do, and it's going to diverge further once we do the restyling anyhow.

File: wagtail/admin/views/pages/listing.py
Comment (gasman) [reply wagtail/admin/views/pages/listing.py:41]: Making the filter form responsible for keeping track of the `search_all` flag feels a bit iffy - that's a detail relating to the `<form>` element that contains both the search and filter forms, and I feel like we've been tying ourselves in knots by not being rigorous enough about deciding who is responsible for that form, what it contains, and under what circumstances it gets submitted. Alongside this particular bug, this is manifesting itself in various other places:

* Searching _does_ respect search_all, because it's not actually submitting/serialising the form - it's plucking out the search input value and combining that with the existing URL querystring, which means that searches don't reflect any changes to filters that have just been made but not submitted
* the need to scatter `is_searching or is_filtering` conditions through the table rendering - it would be neater to treat filtering as a kind of search, but there's code elsewhere in the view that interprets `is_searching` as "send this to a search backend"

I notice that the designs don't have any explicit submit buttons, which might imply that filters are intended to take effect immediately on click/edit, much like search does - this would probably go some way to tightening up this fuzziness. Will ponder some more...

Comment (laymonage) [reply wagtail/admin/views/pages/listing.py:41]: I noticed that the SwapController code already serializes the form **if you use the `submit` method** (not the `search` method), so using `submitLazy` instead of `searchLazy` in the `data-action`, together with my hack above, search+filtering will work with AJAX and `search_all` is retained.

Comment (laymonage) [reply wagtail/admin/views/pages/listing.py:41]: Although that does not clear the `p` param – can probably just copy-paste that from the `search` method. At which point I'm not sure if we'll still have any use for the `search` method instead of the `submit`, provided we use the controller with `SearchForm`...

File: wagtail/admin/templates/wagtailadmin/pages/page_listing_header.html
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/pages/page_listing_header.html:51]: I think we can remove this, it was mostly added to handle the grid layout when the filters are on the right side of the listing.

It also causes alignment issues for checkboxes:

## with the wrapper

image


## without

image

File: wagtail/admin/views/pages/listing.py
Comment (laymonage) [inline wagtail/admin/views/pages/listing.py:38]: I just noticed that if the project has two similarly named models in different apps, e.g. a `Post` (non-page) model in app A and a `Post` page model in app B, the filter could return both. We should also filter by the app label.

I think this is easier with the built-in `ContentType.objects.get_for_models`, which will also make use of the manager's cache when possible. I'll add this as I merge.

Not adding a test yet though, as that would require adding a model without the `Page` suffix in the test models, and I don't have time for it now 😂

Edit: it's harder than I thought, since the filter expects a queryset. I think I'll leave this to be handled in a future PR to unblock things for now.

**Review by laymonage — APPROVED** [review]
I'll approve and merge this to unblock things for now – it's mostly OK, just a few bits that I think we can handle in a separate PR. The bits like the `is_filtering` property is nice and I'd like to apply the same in my PR. Thanks!

**Pull request merged**
